crates/pico/src/database.rs

```diff
@@ -0,0 +1,173 @@
+                    inner_fn,
+                },
+            );
+            self.values.put(node_id, value);
+            time_calculated
+        };
+
+        self.register_dependency(node_id, time_calculated);
+        node_id
+    }
+
+    fn any_dependency_changed(&mut self, node_id: &NodeId) -> bool {
+        let dependencies = self
+            .nodes
+            .get(node_id)
+            .expect("node should exist. This is indicative of a bug in Isograph.")
+            .dependencies
+            .iter()
+            .filter_map(|dep| {
+                if dep.time_verified_or_calculated != self.current_epoch {
+                    Some((dep.node_to, dep.time_verified_or_calculated))
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<_>>();
+
+        dependencies
+            .into_iter()
+            .any(|(dep_node_id, time_verified)| match dep_node_id.kind {
+                NodeKind::Source => self.source_node_changed(&dep_node_id, time_verified),
+                NodeKind::Derived => self.derived_node_changed(&dep_node_id),
+            })
+    }
+
+    fn source_node_changed(&mut self, node_id: &NodeId, time_verified: u64) -> bool {
+        match self.sources.get(node_id) {
+            Some(source) => source.time_calculated > time_verified,
+            None => true,
+        }
+    }
+
+    fn derived_node_changed(&mut self, node_id: &NodeId) -> bool {
+        if !self.params.contains(&node_id.param_id) {
+            return true;
+        }
+        let new_value = (self.nodes.get(node_id).unwrap().inner_fn)(self, node_id.param_id);
+        match self.values.get(node_id) {
+            Some(value) => *value != new_value,
+            None => true,
+        }
+    }
+
+    fn call_inner_fn_and_collect_dependencies(
+        &mut self,
+        param_id: ParamId,
+        inner_fn: impl Fn(&mut Database, ParamId) -> Box<dyn DynEq>,
+    ) -> (
+        Box<dyn DynEq>,  /* value */
+        Vec<Dependency>, /* dependencies */
+        u64,             /* time_calculated */
+    ) {
+        self.dependency_stack.push(vec![]);
+        let value = inner_fn(self, param_id);
+        let registred_dependencies = self
+            .dependency_stack
+            .pop()
+            .expect("dependency stack to not be empty. This is indicative of a bug in Isograph.");
+
+        let (dependencies, time_calculated) = registred_dependencies.into_iter().fold(
+            (vec![], 0),
+            |(mut deps, mut max_time_calculated), (time_calculated, dep)| {
+                deps.push(dep);
+                max_time_calculated = std::cmp::max(max_time_calculated, time_calculated);
+                (deps, max_time_calculated)
+            },
+        );
+        (value, dependencies, time_calculated)
+    }
+
+    pub fn register_dependency(&mut self, node_id: NodeId, time_calculated: u64) {
+        if let Some(dependencies) = self.dependency_stack.last_mut() {
```

Comment:
should we panic if none is found?

Reply:
I'll add it just in case. `inner_fn` is always called after `push(vec![])` but it is not guaranteed if we change something accidentally

Reply:
Oh, actually I made it on purpose to avoid unnecessary register call on a root memoized function. This is the only place where the dependency stack can be empty as all other inner_fn call wrapped in `dependency_stack.push(vec![])`